File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2018/Y18D06.kt
package org.gristle.pdxpuzzles.advent.y2018

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.Coord
import org.gristle.pdxpuzzles.utilities.objects.MutableGrid
import org.gristle.pdxpuzzles.utilities.objects.getBounds
import org.gristle.pdxpuzzles.utilities.parsing.getInts
import kotlin.math.max

class Y18D06(input: String) : Day {

    // coordinates from input, but offset so top left coordinate is normalized to (0,0) 
    private val coordinates: List<Coord>

    // width and height of starfield
    private val width: Int
    private val height: Int

    init {

        // "raw" coordinates before offset parsed from input
        val offsetCoordinates = input
            .getInts()
            .chunked(2) { (x, y) -> Coord(x, y) }
            .toList()

        val bounds = offsetCoordinates.getBounds()

        // calculate the offset
        val offset = Coord(-bounds.first.first, -bounds.second.first)

        // return offset coordinates
        coordinates = offsetCoordinates.map { it + offset }

        // return the width and height of the starfield
        width = bounds.first.last - bounds.first.first + 1
        height = bounds.second.last - bounds.second.first + 1
    }

    override fun part1(): Int {

        // array to track the closest coordinate to every space within the starfield. '-1' means unexplored.
        // '-2' means equally distant to at least two coordinates. Any other value is the index of the closest
        // coordinate.
        val space = MutableGrid(width, height) { -1 }

        // seed the space with the original coordinates
        coordinates.forEachIndexed { index, pos -> space[pos] = index }

        // helper for below ring function
        val sides = listOf(
            Coord(0, -1) to Coord(1, 1),
            Coord(1, 0) to Coord(-1, 1),
            Coord(0, 1) to Coord(-1, -1),
            Coord(-1, 0) to Coord(1, -1),
        )

        // function returns all coordinates that are distance t away from the coordinate
        fun Coord.ring(t: Int) = sides
            .flatMap { (initial, direction) ->
                (2..t).runningFold(this + Coord(initial.x * t, initial.y * t)) { pos, _ -> pos + direction }
            }

        // for each distance unit assign unassigned spaces to the closest coordinate 
        timeLoop@ for (distance in 1 until max(width, height)) {

            // track any assignments because if another coordinate also reaches the space at the same distance then
            // the assignment must be undone
            val distanceLog = mutableSetOf<Coord>()

            coordinates.forEachIndexed { index, pos -> // for each coordinate...
                pos
                    .ring(distance) // ...get the spaces reached at that distance
                    .filter { space.validCoord(it) } // remove spaces not in starfield (infinite space)
                    .forEach { ringPos ->
                        when (space[ringPos]) {
                            -1 -> { // when that space is unoccupied...
                                space[ringPos] = index // ...assign the index to it...
                                distanceLog.add(ringPos) // ...and add the assignment to the log
                            }

                            -2 -> {} // when a conflict has already been found, do nothing

                            // when an assignment has already been made at that space, undo it and put '-2' in its
                            // place to denote a conflict
                            else -> if (ringPos in distanceLog) space[ringPos] = -2
                        }
                    }
            }

            // break early if no new assignments were made
            if (distanceLog.isEmpty()) break@timeLoop
        }

        // 'infinite' stores those values that reach the edge of known space, and thus should not be counted
        val infinite = space.row(0).toSet() +
                space.row(space.height - 1) +
                space.column(0) +
                space.column(space.width - 1)

        return space // look at each space in the starfield
            .filter { it !in infinite } // remove those spaces assigned to 'infinite' coordinates
            .groupingBy { it }
            .eachCount() // count occurrences of the remainder
            .maxOf { (_, count) -> count } // return the largest 
    }

    // recreate indices of the starfield and count if the sum of all coordinate distances to that spot is less than
    // 10,000
    override fun part2() = (0 until width * height)
        .count { index -> coordinates.sumOf { it.manhattanDistance(Coord.fromIndex(index, width)) } < 10_000 }
}

fun main() = Day.runDay(Y18D06::class)

//    Class creation: 6ms
//    Part 1: 5365 (173ms)
//    Part 2: 42513 (50ms)
//    Total time: 230ms

File: rust/advent/src/bin/2018/y18d06.rs
use std::cmp::max;

use advent::utilities::get_input::get_input;
use itertools::Itertools;
use rustc_hash::FxHashSet;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::{coord::{Coord2, Coord2U}, grid::Grid2, stopwatch::{ReportDuration, Stopwatch}}};

type Input = (Vec<Pos>, usize, usize);
type Output = usize;
type Pos = Coord2;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(18, 6).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse_input(input: &str) -> Input {
    // "raw" coordinates before offset parsed from input
    let offset_pos: Vec<Pos> = input.get_numbers()
        .chunks(2)
        .into_iter()
        .map(|mut chunk| Pos::new2d(chunk.next().unwrap(), chunk.next().unwrap()))
        .collect();

    let mut x_min = i64::MAX;
    let mut y_min = i64::MAX;
    let mut x_max= i64::MIN;
    let mut y_max= i64::MIN;

    for &pos in offset_pos.iter() {
        let x = pos.x();
        let y = pos.y();
        if x < x_min {
            x_min = x;
        } else if x > x_max {
            x_max = x;
        }
        if y < y_min {
            y_min = y;
        } else if y > y_max {
            y_max = y;
        }
    }

    let offset = Pos::new2d(x_min, y_min);
    
    let positions = offset_pos.into_iter()
        .map(|pos| pos - offset)
        .collect();

    let width = (x_max - x_min + 1) as usize;
    let height = (y_max - y_min + 1) as usize;
    
    (positions, width, height)
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
enum Space {
    Unexplored,
    Equidistant,
    Mapped(usize),
}

fn part1(input: &Input) -> Output {
    let (positions, width, height) = input;
    let mut space = Grid2::new2d(vec![Space::Unexplored; width * height], *width).unwrap();
    for (index, &pos) in positions.iter().enumerate() {
        let pos = Coord2U::try_from(pos).unwrap();
        space[pos] = Space::Mapped(index);
    }

    let sides = [
        (Pos::new2d(0, -1), Pos::new2d(1, 1)),
        (Pos::new2d(1, 0), Pos::new2d(-1, 1)),
        (Pos::new2d(0, 1), Pos::new2d(-1, -1)),
        (Pos::new2d(-1, 0), Pos::new2d(1, -1)),
    ];

    let ring = |pos: Pos, t: i64| {
        sides.iter()
            .copied()
            .flat_map(move |(initial, dir)| {
                let initial = pos + Pos::new2d(initial.x() * t, initial.y() * t);
                let it = (2..=t).scan(initial, move |acc, _| {
                    *acc = *acc + dir;
                    Some(*acc)
                });
                std::iter::once(initial).chain(it)
            })
    };

    // for each distance unit assign unassigned spaces to the closest coordinate 
    for distance in 1..max(*width, *height) {
        // track any assignments because if another coordinate also reaches the space at the same distance then
        // the assignment must be undone
        let mut distance_log = FxHashSet::default();

        for (index, &pos) in positions.iter().enumerate() {
            for ring_pos in ring(pos, distance as i64) {
                let ring_pos = Coord2U::try_from(ring_pos)
                    .unwrap_or(Coord2U::new2d(99_999, 99_999));
                match space.get(ring_pos) {
                    None | Some(Space::Equidistant) => {},
                    Some(Space::Unexplored) => {
                        space[ring_pos] = Space::Mapped(index);
                        distance_log.insert(ring_pos);
                    },
                    _ => {
                        if distance_log.contains(&ring_pos) {
                            space[ring_pos] = Space::Equidistant;
                        }
                    },
                }
            }
        }
        if distance_log.is_empty() {
            break;
        }
    }

    let infinite: FxHashSet<_> = 
        space.row(0).unwrap()
            .chain(space.row(*height - 1).unwrap())
            .chain(space.column(0).unwrap())
            .chain(space.column(*width - 1).unwrap())
            .collect();

    space.iter()
        .filter(|&it| !infinite.contains(it))
        .counts()
        .iter()
        .map(|(_, &count)| count)
        .max()
        .unwrap()   
}

fn part2(input: &Input) -> Output {
    let (positions, width, height) = input;
    (0..width * height)
        .filter(|&idx| {
            let sum: usize = positions.iter()
                .map(|pos| {
                    pos.manhattan_distance(Pos::from_index(idx, *width).unwrap())
                })
                .sum();
            sum < 10_000
        })
        .count()
}

#[test]
fn default() {
    let input = get_input(18, 6).unwrap();
    let input = parse_input(&input);
    assert_eq!(5365, part1(&input));
    assert_eq!(42513, part2(&input));
}

// Input parsed (19μs)
// 1. 5365 (13ms)
// 2. 42513 (4ms)
// Total: 18ms
